stdlib/public/Python/Array+NumPy.swift

```diff
@@ -0,0 +1,79 @@
+//===-- Array+NumPy.swift -------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines bridging from `numpy.ndarray` to `Array`.
+//
+//===----------------------------------------------------------------------===//
+
+public extension Array {
+  init?(numpyArray: PyValue) {
```

Comment:
We could add `init?(numpyArray: PyValue, flattened: Bool = false)` so the code could be reused in the TensorFlow module?

Reply:
I can't comment on your comment above about reusing the `Array(numpyArray:)` initializer for `ShapedArray` for some reason.

As mentioned on the issue:
> Something to consider is how to handle bridging higher dimensional `numpy.ndarray` instances to `Array`. It makes sense for the bridging initializer to fail when the `numpy.ndarray` instance is not 1-D.

I think implicit flattening when converting to `Array` is bad.
Adding a `flattened: Bool = false` argument would make things explicit, but semantically flattening is not fundamentally tied to conversion. Users should be able to call the `init?(numpyArray:)` initializer without thinking about flattening.

If anything, there should be separate initializers:
* `init?(numpyArray: PyValue)` (for simple conversion)
* `init?(flatteningNumpyArray: PyValue)` (for flattening conversion)
This leads us back to the original question about implicit flattening for `Array(numpyArray:)`, however.

One argument against implicit flattening is the existence of [`Array1D`](https://github.com/google/swift/blob/5ff63acc1c174516f2fea5bc77c67f0595cb8b0d/stdlib/public/TensorFlow/RankedArray.swift.gyb#L27), which is a typealias of `Array`.
If `Array(numpyArray:)` doesn't fail when converting 2D+ `ndarray` instances, code like  `Array1D(numpyArray: np.ones.call(with: [10, 10, 10])` would counterintuitively succeed.

In any case, I discussed a protocol-based reimplementation of `ndarray` conversion with @rxwei and I'll push it soon.

Reply:
> I think implicit flattening when converting to Array is bad.

I agree! The _good_ solution would be to be able to convert NumPy arrays as _ranked_ Swift arrays and to add a initializer to `ShapedArray` which supported ranked Swift arrays. This should be the long term solution I think.

> In any case, I discussed a protocol-based reimplementation of ndarray conversion with @rxwei and I'll push it soon.

Looking forward for it :)

Reply:
@pvieito Adding initializers to ranked arrays would be very nice, however, we've decided to remove ranked arrays/tensors for various reasons. I'm going to start a discussion on the mailing list.

Reply:
> By "_ranked_ Swift arrays", I believe @pvieito was actually referring to nested `Array` types like [[Float]], [[[Float]]], etc.

Yes, exactly. I know it would be difficult to implement (maybe impossible?) but probably would be the cleanest way to do the conversion from NumPy to Tensor through Swift `Array`.

Reply:
Please check out the new protocol-based implementation in 035acb6e142bd34ee5baa52e8cf8bb623fa337ba!

> I know it would be difficult to implement (maybe impossible?) but probably would be the cleanest way to do the conversion from NumPy to Tensor through Swift `Array`.

Doing `ndarray -> Array -> Tensor` is less efficient than `ndarray -> Tensor` because there's an extra copy. Reducing the number of copies is more important than code dupe.

Reply:
> Doing ndarray -> Array -> Tensor is less efficient than ndarray -> Tensor because there's an extra copy. Reducing the number of copies is more important than code dupe.

Ok, that solves my main concern. Also, the new protocol-based implementation looks a lot cleaner :)